core/modules/filter/filter.pages.inc

```diff
@@ -238,20 +279,57 @@ function _filter_image_library_ajax() {
  * Submit handler for filter_format_editor_image_form().
  */
 function filter_format_editor_image_form_submit($form, &$form_state) {
-  // Convert any uploaded files from the FID values to the src attribute.
-  if (!empty($form_state['values']['fid'])) {
+  // Determine the file either from fid (when upload is enabled) or from src
+  // attribute (if image upload in editor is disabled).
+  $file = NULL;
+  $fid = 0;
+  $attributes = $form_state['values']['attributes'] ?? array();
+  if (!empty($form_state['values']['fid']) && empty($attributes['src'])) {
+    $file = file_load($form_state['values']['fid']);
     $fid = $form_state['values']['fid'];
-    $file = file_load($fid);
-
+    unset($form_state['values']['fid']);
+  }
+  elseif (!empty($attributes['src'])) {
+    $fid = _filter_get_file_id($attributes['src']);
+    if ($fid) {
+      $file = file_load($fid);
+      unset($attributes['src']);
+    }
+  }
+  if (!empty($file)) {
     // Try to make a local path if possible for better portability.
     $absolute_path = parse_url($GLOBALS['base_url'], PHP_URL_PATH) . '/';
-    $url = file_create_url($file->uri);
+
+    if (empty($form_state['values']['image_style'])) {
+      $url = file_create_url($file->uri);
+      if (empty($attributes['width']) && empty($attributes['height'])) {
+        $attributes['width'] = $file->width;
+        $attributes['height'] = $file->height;
+      }
+    }
+    else {
+      $style_name = $form_state['values']['image_style'];
+      $url = image_style_url($style_name, $file->uri);
+      $dimensions = array('width' => '', 'height' => '');
+      // Although the editor could determine dimensions, this would not work
+      // with SVG.
+      if (!empty($file->width)) {
+        $dimensions = array(
+          'width' => $file->width,
+          'height' => $file->height,
+        );
+        image_style_transform_dimensions($style_name, $dimensions);
+      }
+      $attributes['width'] = $dimensions['width'];
+      $attributes['height'] = $dimensions['height'];
```

Comment:
```suggestion
      if ($form['image_style']['#default_value'] != $style_name) {
        $attributes['width'] = $dimensions['width'];
        $attributes['height'] = $dimensions['height'];
      };
```

Likewise, I think only when the selected image style is different from that already set should we override the dimensions. Otherwise, every time something else is changed it will also reset the dimensions.

Reply:
Overriding per style already happens (AFAIK) on submit, when a style is set.

Reply:
> Likewise, I think only when the selected image style is different from that already set should we override the dimensions.

This problem was more tricky to solve. If I blank out the height/width so that the original image can use the original dimensions, the image style scenario doesn't have previous dimensions to reference. I added hidden (`#type = value`) elements to hold onto that information so we have it if we need to restore it on submit.

Reply:
I think this is fixed now, though it took extra form elements to handle.